server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java

```diff
@@ -468,13 +467,7 @@ public RetentionLeases loadRetentionLeases(final Path path) throws IOException {
         synchronized (retentionLeasePersistenceLock) {
             retentionLeases = RetentionLeases.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, path);
         }
-        return emptyIfNull(retentionLeases);
-    }
-
-    @UpdateForV9(owner = UpdateForV9.Owner.DISTRIBUTED_INDEXING)
-    private static RetentionLeases emptyIfNull(RetentionLeases retentionLeases) {
-        // we expect never to see a null in 8.x, so adjust this to throw an exception from v9 onwards.
-        return retentionLeases == null ? RetentionLeases.EMPTY : retentionLeases;
+        return Objects.requireNonNull(retentionLeases, "Unable to load retention leases from the state file " + path);
```

Comment:
Looks like we still can get null here when recovering a shard:
```
[2024-10-08T15:28:52,946][WARN ][o.e.i.c.IndicesClusterStateService] [node_s4] [test][1] marking and sending shard failed due to [failed recovery]
org.elasticsearch.indices.recovery.RecoveryFailedException: [test][1]: Recovery failed from {node_s7}{2jMKcZYtQB2HVsjZGx7Glw}{lg7gYBZlTxKGL8Ppe_Uigw}{node_s7}{127.0.0.1}{127.0.0.1:13427}{d}{9.0.0}{8000099-8517000} into {node_s4}{VRMYHpHNS1--tSq3Poe7Ng}{0UGPlLJBR6WxP3RljvqB-A}{node_s4}{127.0.0.1}{127.0.0.1:13421}{cdfhilrstw}{9.0.0}{8000099-8517000} (failed to prepare shard for recovery)
	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService.lambda$doRecovery$1(PeerRecoveryTargetService.java:297) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:188) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:193) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75) ~[main/:?]
	at org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onFailure(ActionListenerImplementations.java:346) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75) ~[main/:?]
	at org.elasticsearch.action.ActionListener$3.onFailure(ActionListener.java:410) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations$NotifyOnceActionListener.onFailure(ActionListenerImplementations.java:376) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75) ~[main/:?]
	at org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:93) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:184) ~[main/:?]
	at org.elasticsearch.index.CompositeIndexEventListener.callListeners(CompositeIndexEventListener.java:291) ~[main/:?]
	at org.elasticsearch.index.CompositeIndexEventListener.iterateBeforeIndexShardRecovery(CompositeIndexEventListener.java:300) ~[main/:?]
	at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:327) ~[main/:?]
	at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1800) ~[main/:?]
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:452) ~[main/:?]
	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService.doRecovery(PeerRecoveryTargetService.java:305) ~[main/:?]
	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$RecoveryRunner.doRun(PeerRecoveryTargetService.java:740) ~[main/:?]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:992) ~[main/:?]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27) ~[main/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1570) ~[?:?]
Caused by: java.lang.NullPointerException: Unable to load retention leases from the state file /home/artem/IdeaProjects/elasticsearch-stateless/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.cluster.routing.ShardRoutingRoleIT_7D4B552262BB29D7-001/tempDir-002/node_s4/indices/ac9ZuaecQrmCciVRDFPmZg/1
	at java.util.Objects.requireNonNull(Objects.java:246) ~[?:?]
	at org.elasticsearch.index.seqno.ReplicationTracker.loadRetentionLeases(ReplicationTracker.java:470) ~[main/:?]
	at org.elasticsearch.index.shard.IndexShard.loadRetentionLeases(IndexShard.java:2867) ~[main/:?]
	at org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:2144) ~[main/:?]
	at org.elasticsearch.index.shard.IndexShard.openEngineAndSkipTranslogRecovery(IndexShard.java:2130) ~[main/:?]
	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService.lambda$doRecovery$2(PeerRecoveryTargetService.java:311) ~[main/:?]
	at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:91) ~[main/:?]
	... 13 more
[2024-10-08T15:28:52,947][WARN ][o.e.i.c.IndicesClusterStateService] [node_s8] [test][0] marking and sending shard failed due to [failed recovery]
```

Reply:
It fails on opening the engine during a peer recovery
```
  1> Caused by: java.lang.NullPointerException: Unable to load retention leases from the state file /home/artem/IdeaProjects/elasticsearch-stateless/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.cluster.routing.ShardRoutingRoleIT_161392420F42C0CA-001/tempDir-002/node_s5/indices/DwwkWhkgS-OX63SQzN6PDg/0
  1> 	at java.util.Objects.requireNonNull(Objects.java:246) ~[?:?]
  1> 	at org.elasticsearch.index.seqno.ReplicationTracker.loadRetentionLeases(ReplicationTracker.java:472) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.loadRetentionLeases(IndexShard.java:2867) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:2144) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.openEngineAndSkipTranslogRecovery(IndexShard.java:2130) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService.lambda$doRecovery$2(PeerRecoveryTargetService.java:311) ~[main/:?]
  1> 	at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:91) ~[main/:?]
  1> 	... 13 more
  ```

Reply:
```
  1> [2024-10-09T16:42:17,973][WARN ][o.e.g.MetadataStateFormat] [node_s5] Unable to resolve /home/artem/IdeaProjects/elasticsearch-stateless/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.cluster.routing.ShardRoutingRoleIT_161392420F42C0CA-001/tempDir-002/node_s5/indices/DwwkWhkgS-OX63SQzN6PDg/0/_state
1> [2024-10-09T16:42:17,975][INFO ][o.e.g.MetadataStateFormat] [node_s5] Reading maxGenerationId = -1, from [/home/artem/IdeaProjects/elasticsearch-stateless/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.cluster.routing.ShardRoutingRoleIT_161392420F42C0CA-001/tempDir-002/node_s5/indices/DwwkWhkgS-OX63SQzN6PDg/0]
  1> [2024-10-09T16:42:17,977][INFO ][o.e.i.s.ReplicationTracker] [node_s5] [test][0] Loading state files: [] by generation: -1 from [/home/artem/IdeaProjects/elasticsearch-stateless/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.cluster.routing.ShardRoutingRoleIT_161392420F42C0CA-001/tempDir-002/node_s5/indices/DwwkWhkgS-OX63SQzN6PDg/0]
  1> [2024-10-09T16:42:17,977][INFO ][o.e.i.s.ReplicationTracker] [node_s5] [test][0] Loading retention leases null from /home/artem/IdeaProjects/elasticsearch-stateless/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.cluster.routing.ShardRoutingRoleIT_161392420F42C0CA-001/tempDir-002/node_s5/indices/DwwkWhkgS-OX63SQzN6PDg/0
 ```

Reply:
I believe it this fails for in `ShardRoutingRoleIT` and `SearchableSnapshotsIntegTests`